CIP-XXXX/README.md

```diff
@@ -0,0 +1,69 @@
+
+Additionally, this is not a problem that can easily be solved with better indexing tools for Cardano. This is because in the proxy native script contract use-case, it's possible that the proxy contract is unique for each individual user (some parameterized Plutus contract is calculated and then added as the requirement of the native script). Therefore, this feature benefits from being supported at the address level.
+
+Lastly, the fact that addresses contain just their script hash without information about which language they use also leads to complexity in ledger rules and code of SDKs for Cardano as they cannot process any script language specific behavior until after the scripts are later provided as part of the witness. Knowing the language ahead of time allows the ledger to protect users from mistakes such as creating a Plutus V1 utxo entry with inline datum (which effectively makes it unspendable)
+
+# Specification
+
+All addresses in Cardano currently start with a header that indicates which address type they are. Notably, these can be seen in [CIP19](../CIP-0019/README.md).
+
+Note there are two ways we can support our modification to the address type:
+
+1) Reserve a new address type as part of the header nibble. This is not great because it means we have to reserve many new address types (a new alternative for every case where the old address format contains a script hash)
+2) Add a new column to existing address types that contain scripts. This doesn't work for pointer addresses because they are variable-width so it isn't possible to differentiate between the original address format and this annotated format (note that pointer addresses are almost never used on chain, so this is an acceptable limitation). This does, however, limit the kinds of modifications we could make to these address types in the future as we would need to make sure the length continues to uniquely identify the content.
+
+In this CIP, we propose using option (2) and extend the CIP19 table as follows:
+
+Header type (`t t t t . . . .`) | script namespace* | Payment Part     | script namespace* | Delegation Part
+---                             | ---               | ---              | ---               |---
+(0) `0000....`                  | ø                 | `PaymentKeyHash` | ø                 | `StakeKeyHash`
+(1) `0001....`                  | 0-255             | `ScriptHash`     | ø                 | `StakeKeyHash`
```

Comment:
Note that the size of the namespace is not defined in the Cardano binary spec, so I just made it one byte. At the current rate (3 namespaces in 2 years since the Shelley release), we have enough for 84 years which seems decent.

If we wanted to optimize for address length, we have 3 optimization targets:
1. Lowest amount of size on-chain (every bit counts). Probably we should avoid weird length addresses though
2. Smallest hex representation (0xF vs 0xFF i.e. 0-16 vs 0-255). 16 seems way too small though
3. Smallest bech32 representation (0-32 vs 0-1024 since bech32 takes 5 bit chunks). Using this also leads to weird length addresses though